File: rust/src/tokenizer.rs
use porter_stemmer::stem;
use regex::Regex;
pub struct Tokenizer {
}

impl Tokenizer {
    pub fn new() -> Tokenizer {
        Tokenizer { }
    }
    pub fn tokenize(&self, token: &str) -> Vec<String> {
        let pattern = Regex::new(r"[^a-zA-Z0-9]+").unwrap();
        pattern
            .split(token)
            .map(|s| s.to_lowercase())
            .filter(|s| !self.stop_words(s))
            .map(|s| self.porter_stemmer(&s))
            .collect::<Vec<String>>()
    }

    pub fn porter_stemmer(&self, token: &str) -> String {
        stem(token)
    }

}

File: python/src/tokenizer.py
# ! also stemmer and normalizer and remove stop words & punctuations
# * NTLK: https://www.nltk.org/
# * classification, tokenization, stemming, tagging, parsing, and semantic reasoning, wrappers for industrial-strength NLP libraries

import nltk # type: ignore

class Tokenizer:
    def __init__(self):
        self.porter_stemmer = nltk.stem.PorterStemmer()

    def tokenize(self, text: str) -> list[str]:

        #! Split the input text into tokens, using Porter Stemmer.
        
        #! Get rid of prefixes and suffixes, so that each word can
        #! be a token and be the same token as other similar words.

        #! Note: We can't convert tokens into a set, since we would
        #! lose count of how many times each token shows up. Therefore
        #! we are doing tokenization during both indexing and querying.

        #! Tokenize the text with regex, so that we only get alphanumeric
        tokens = nltk.regexp_tokenize(text, r'[A-Za-z0-9]+')
        #! Return a list of stemmed tokens
        return [self.porter_stemmer.stem(token, to_lowercase=True) for token in tokens]

#! Testing the tokenizer
if __name__ == "__main__":
    tokenizer = Tokenizer()
    assert tokenizer.tokenize("This is a simple text.") == [
        "thi", "is", "a", "simpl", "text"]
    assert tokenizer.tokenize("THIS IS  A simpler TEXT WITH MORE WORDS.") == [
        "thi", "is", "a", "simpler", "text", "with", "more", "word"]
    assert tokenizer.tokenize("iftekhar ahmed") == [
        "iftekhar", "ahm"]
